src/scale_impls/encode.rs

```diff
@@ -209,21 +218,48 @@ fn encode_composite<T>(
     Err(original_error)
 }
 
-// skip into the target type past any newtype wrapper like things:
-fn find_single_entry_with_same_repr(type_id: u32, types: &PortableRegistry) -> u32 {
-    let Some(ty) = types.resolve(type_id) else { return type_id };
-    match &ty.type_def {
-        TypeDef::Tuple(tuple) if tuple.fields.len() == 1 => {
-            find_single_entry_with_same_repr(tuple.fields[0].id, types)
-        }
-        TypeDef::Composite(composite) if composite.fields.len() == 1 => {
-            find_single_entry_with_same_repr(composite.fields[0].ty.id, types)
-        }
-        TypeDef::Array(arr) if arr.len == 1 => {
-            find_single_entry_with_same_repr(arr.type_param.id, types)
-        }
-        _ => type_id,
+/// Skip into the target type past any newtype wrapper like things.
+/// Also returns a bool indicating whether we skipped into something or not.
+/// This bool is true when the returned id is different from the id that was passed in.
+fn find_single_entry_with_same_repr<'a, R: TypeResolver>(
+    type_id: &'a R::TypeId,
+    types: &'a R,
+) -> (&'a R::TypeId, bool) {
+    fn do_find<'a, R: TypeResolver>(
+        type_id: &'a R::TypeId,
+        types: &'a R,
+        type_id_has_changed: bool,
+    ) -> (&'a R::TypeId, bool) {
+        let return_unchanged = (type_id, type_id_has_changed);
+        let visitor = scale_type_resolver::visitor::new((), |_, _| return_unchanged)
+            .visit_tuple(|_, fields| {
+                if fields.len() == 1 {
+                    let ty = fields.next().expect("has 1 item; qed;");
+                    do_find(ty, types, true)
```

Comment:
is it possible to re-write this without recursion? Unless, I'm really slow it looks like recursion to me

Reply:
Yes, this uses recursion, but I don't think we can avoid it. The implementation before this PR had the same recursion going on. If we write a manual visitor implementation instead of using the `ConcreteResolvedTypeVisitor` we could get rid of the `do_find` wrapper function by passing the visitor itself to each new visit, but this is essentially the same recursion.

Reply:
It should never recurse very much, so I wouldn't be very worried about it myself :) It probably isn't too hard to re-write into a loop but might look uglier!